pkg/autodiscovery/providers/kube_endpoints.go

```diff
@@ -30,9 +31,10 @@ const (
 
 // KubeEndpointsConfigProvider implements the ConfigProvider interface for the apiserver.
 type KubeEndpointsConfigProvider struct {
-	serviceLister   listersv1.ServiceLister
-	endpointsLister listersv1.EndpointsLister
-	upToDate        bool
+	serviceLister      listersv1.ServiceLister
+	endpointsLister    listersv1.EndpointsLister
+	upToDate           bool
+	monitoredEndpoints map[string]bool
```

Comment:
You should add a mutex to protect the `map` since `Collect()`, `invalidateIfChangedService()` and `invalidateIfChangedEndpoints` is updating it in different goroutines.

Reply:
indeed, good catch 👍